zwave_js_server/client.py

```diff
@@ -199,181 +121,122 @@ async def async_send_json_message(self, message: Dict[str, Any]) -> None:
+            raise InvalidState("Not connected when start listening")
 
-                if msg.type != WSMsgType.TEXT:
-                    disconnect_warn = "Received non-Text message: {}".format(msg.type)
-                    break
+        assert self.client
 
-                try:
-                    if len(msg.data) > SIZE_PARSE_JSON_EXECUTOR:
-                        msg = await loop.run_in_executor(None, msg.json)
-                    else:
-                        msg = msg.json()
-                except ValueError:
-                    disconnect_warn = "Received invalid JSON."
-                    break
+        while not self.client.closed:
+            try:
+                msg = await self.client.receive()
+            except asyncio.CancelledError:
+                break
 
-                if self._logger.isEnabledFor(logging.DEBUG):
-                    self._logger.debug("Received message:\n%s\n", pprint.pformat(msg))
+            if msg.type in (WSMsgType.CLOSED, WSMsgType.CLOSING):
+                break
 
-                msg_ = cast(dict, msg)
-                try:
-                    self.async_handle_message(msg_)
+            if msg.type == WSMsgType.CLOSE:
+                raise ConnectionClosed()
 
-                except InvalidState as err:
-                    disconnect_warn = f"Invalid state: {err}"
-                    await client.close()
-                    break
+            if msg.type == WSMsgType.ERROR:
+                raise ConnectionFailed()
 
-                except Exception:  # pylint: disable=broad-except
-                    self._logger.exception("Unexpected error handling %s", msg)
-                    break
+            if msg.type != WSMsgType.TEXT:
+                raise InvalidMessage(f"Received non-Text message: {msg.type}")
 
-        except client_exceptions.WSServerHandshakeError as err:
-            self._logger.warning("Unable to connect: %s", err)
+            try:
+                if len(msg.data) > SIZE_PARSE_JSON_EXECUTOR:
+                    msg = await self._loop.run_in_executor(None, msg.json)
+                else:
+                    msg = msg.json()
+            except ValueError as err:
+                raise InvalidMessage("Received invalid JSON.") from err
 
-        except client_exceptions.ClientError as err:
-            self._logger.warning("Unable to connect: %s", err)
+            if self._logger.isEnabledFor(logging.DEBUG):
+                self._logger.debug("Received message:\n%s\n", pprint.pformat(msg))
 
-        except asyncio.CancelledError:
-            pass
+            msg_ = cast(dict, msg)
 
-        finally:
-            if disconnect_warn is None:
-                self._logger.info("Connection closed")
-            else:
-                self._logger.warning("Connection closed: %s", disconnect_warn)
+            try:
+                self.async_handle_message(msg_)
+            except InvalidState as err:
+                await self._close(err)
```

Comment:
I wonder if we shoult not close here, but just let the exception be raised? We handle the exception during `connect` now, and if the connect hasn't finished there's no point in closing. It's also the repsonsibility of the library user now to disconnect.

Reply:
A reason to still close here is to cancel the futures waiting for a received message.

Reply:
We probably need to close the connection then for all exception raised in listen.

Reply:
I've simplified the error propagation between listen and connect now.